daemon/src/maker_inc_connections_actor.rs

```diff
@@ -1,95 +1,225 @@
-                        maker_inc_connections_actor::Command::OutProtocolMsg { taker_id, msg } => {
-                            let conn = write_connections.get(&taker_id).expect("no connection to taker_id");
-                            conn.send(wire::MakerToTaker::Protocol(msg)).unwrap();
-                        }
-                    }
+pub struct BroadcastOrder(pub Option<Order>);
+impl Message for BroadcastOrder {
+    type Result = anyhow::Result<()>;
+}
+
+pub struct SendOrder {
+    pub order: Option<Order>,
+    pub taker_id: TakerId,
+}
+impl Message for SendOrder {
+    type Result = anyhow::Result<()>;
+}
+
+pub struct NotifyInvalidOrderId {
+    pub id: OrderId,
+    pub taker_id: TakerId,
+}
+impl Message for NotifyInvalidOrderId {
+    type Result = anyhow::Result<()>;
+}
+
+pub struct NotifyOrderAccepted {
+    pub id: OrderId,
+    pub taker_id: TakerId,
+}
+impl Message for NotifyOrderAccepted {
+    type Result = anyhow::Result<()>;
+}
+
+pub struct OutProtocolMsg {
+    pub taker_id: TakerId,
+    pub msg: SetupMsg,
+}
+impl Message for OutProtocolMsg {
+    type Result = anyhow::Result<()>;
+}
+
+type MakerToTakerConnection = mpsc::UnboundedSender<wire::MakerToTaker>;
+
+pub struct MakerIncConnectionsActor {
+    write_connections: Arc<Mutex<HashMap<TakerId, MakerToTakerConnection>>>,
+}
+impl Actor for MakerIncConnectionsActor {}
+
+impl MakerIncConnectionsActor {
+    pub fn new(listener: TcpListener, cfd_maker_actor_address: Address<MakerCfdActor>) -> Self {
+        let write_connections = Arc::new(Mutex::new(HashMap::<
+            TakerId,
+            mpsc::UnboundedSender<wire::MakerToTaker>,
+        >::new()));
+
+        {
+            let write_connections = write_connections.clone();
+            tokio::spawn(async move {
+                loop {
+                    if let Ok((socket, remote_addr)) = listener.accept().await {
+                        println!("Connected to {}", remote_addr);
+                        let taker_id = TakerId::default();
+                        let (read, write) = socket.into_split();
+
+                        let in_taker_actor =
+                            in_taker_messages(read, cfd_maker_actor_address.clone(), taker_id);
+                        let (out_message_actor, out_message_actor_inbox) =
+                            send_wire_message_actor::new(write);
+
+                        tokio::spawn(in_taker_actor);
+                        tokio::spawn(out_message_actor);
+
+                        cfd_maker_actor_address
+                            .do_send_async(maker_cfd_actor::NewTakerOnline { id: taker_id })
+                            .await
+                            .unwrap();
+
+                        write_connections
+                            .lock()
+                            .await
+                            .insert(taker_id, out_message_actor_inbox);
+                    };
                 }
-            }
+            });
```

Comment:
We could do this outside of this actor and pass _this_ actor's `Address` into the loop, then we could have an incoming message for passing a new socket to it. That would allow you to avoid the `Arc<Mutex>` because inside `Handler` you have a mutable reference.

Reply:
not sure if I did it exactly the way you suggested, but it did remove the need for `Arc<Mutex>` :)

thanks for the suggestion, I wasn't happy with the previous solution!